### DISCUSSION TITLE
Sorbet-compatible attachment modules?

### DISCUSSION THREAD

**Discussion by emilong** [shrinerb/shrine#670] [Ideas]
When including dynamically instantiated modules such as `include ImageUploader::Attachment.new(:image)`, Sorbet raises [an error #4002](https://sorbet.org/docs/error-reference#4002) that it can't statically analyze the class including the module.

Instead of configuring the `Attachment` module in its constructor, would it be possible to configure it in a DSL?

for example:

```ruby
class MyModel
  include ImageUploader::Attachment
  attachment_name :image
end
```

Or is there some other way to create a Sorbet-friendly way to incorporate Shrine?

(Note this was formerly a [discussion on the discourse forums](https://discourse.shrinerb.com/t/sorbet-friendly-attachment-declarations/617) but it's still relevant)

**Comment by janko** [comment]
This API would not allow you to have two attachment attributes on the same model with different uploaders.

**Reply by emilong** [reply]
Thanks for that feedback - I was just offering an API for illustration purposes, but it's helpful to know why that particular one won't work.

If there are any other options to avoid dynamically instantiated modules and still get all or most of Shrine's features, I'd love to explore that. I'd also be happy to submit a PR if you have any guidance. :pray:

**Reply by janko** [reply]
I don't have any experience with Sorbet, so I don't know what would work best. I think we won't be able to make Shrine's attachment module Sorbet-compatible out-of-the-box, but we can arrive at a solution that anyone can incorporate in their app.

Another idea was to assign the attachment module to a constant, and then possibly annotate it with Sorbet type definitions. Something like:

```rb
ImageAttachment = ImageUploader::Attachment.new(:image)
ImageAttachment.module_eval do
  extend T::Sig
  # ...
end

class MyModel
  include ImageAttachment
end
```

Don't know if Sorbet supports typing signatures of methods that have already been defined. Or maybe just assigning the module to a constant is enough for Sorbet?
